internal/http/docs/basic_auth.go

```diff
@@ -0,0 +1,156 @@
+package docs
+
+import (
+	"crypto/md5"
+	"crypto/sha256"
+	"crypto/subtle"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"golang.org/x/crypto/bcrypt"
+	"golang.org/x/crypto/scrypt"
+
+	"github.com/benthosdev/benthos/v4/internal/docs"
+)
+
+const (
+	scryptN      = 32768
+	scryptR      = 8
+	scryptP      = 1
+	scryptKeyLen = 32
+)
+
+// BasicAuth contains the configuration fields for the HTTP BasicAuth
+type BasicAuth struct {
+	Enabled      bool   `json:"enabled" yaml:"enabled"`
+	Username     string `json:"username" yaml:"username"`
+	PasswordHash string `json:"password_hash" yaml:"password_hash"`
+	Realm        string `json:"realm" yaml:"realm"`
+	Algorithm    string `json:"algorithm" yaml:"algorithm"`
+	Salt         string `json:"salt" yaml:"salt"`
+}
+
+// NewBasicAuth returns a BasicAuth with default values
+func NewBasicAuth() BasicAuth {
+	return BasicAuth{
+		Enabled:      false,
+		Username:     "",
+		PasswordHash: "",
+		Realm:        "restricted",
+		Algorithm:    "sha256",
+		Salt:         "",
+	}
+}
+
+// Validate confirms that the BasicAuth is properly configured
+func (b BasicAuth) Validate() error {
+	if !b.Enabled {
+		return nil
+	}
+
+	if b.Username == "" || b.PasswordHash == "" {
+		return errors.New("both username and password_hash are required")
+	}
+
+	if !(b.Algorithm == "md5" || b.Algorithm == "sha256" || b.Algorithm == "bcrypt" || b.Algorithm == "scrypt") {
+		return errors.New("algorithm should be one of md5, sha256, bcrypt, or scrypt")
+	}
+
+	if b.Algorithm == "scrypt" && b.Salt == "" {
+		return errors.New("salt is required for scrypt")
+	}
+
+	if b.Algorithm == "scrypt" {
+		if _, err := base64.StdEncoding.DecodeString(b.Salt); err != nil {
+			return fmt.Errorf("invalid salt : %w", err)
+		}
+	}
+
+	return nil
+}
+
+// WrapHandler wraps the provided HTTP handler with middleware that enforces
+// BasicAuth if it's enabled.
+func (b BasicAuth) WrapHandler(next http.HandlerFunc) http.HandlerFunc {
+	if !b.Enabled {
+		return next
+	}
+
+	return func(next http.HandlerFunc, auth BasicAuth) http.HandlerFunc {
```

Comment:
What's the value of this additional closure? It doesn't seem like anything passed is dynamic and therefore needs to be passed by value.

Reply:
In and earlier iteration I was closing over some other stuff. I think it should be fine to remove it. Thanks for looking out!